File: mobile_app/lib/services/mqtt_service.dart
import 'package:mqtt_client/mqtt_client.dart';
import 'package:mqtt_client/mqtt_server_client.dart';

class MqttService {
  late MqttServerClient client;
  final String broker = "localhost";
  final int port = 1883;

  Future<void> connect() async {
    client = MqttServerClient(broker, '');
    client.port = port;
    client.logging(on: true);
    client.keepAlivePeriod = 20;

    try {
      await client.connect();
      print("Connected to MQTT broker");
    } catch (e) {
      print("Error connecting to MQTT broker: $e");
      client.disconnect();
    }
  }

  void subscribe(String topic, Function(String) onMessage) {
    client.subscribe(topic, MqttQos.atMostOnce);
    client.updates?.listen((List<MqttReceivedMessage<MqttMessage>> events) {
      final recMessage = events[0].payload as MqttPublishMessage;
      final message =
          MqttPublishPayload.bytesToStringAsString(recMessage.payload.message);
      onMessage(message);
    });
  }

  void publish(String topic, String message) {
    final builder = MqttClientPayloadBuilder();
    builder.addString(message);
    client.publishMessage(topic, MqttQos.atMostOnce, builder.payload!);
  }
}

File: web_app/src/sevices/mqttservice.js
import mqtt from 'mqtt';

class MqttService {
  constructor() {
    this.client = null;
  }

  connect() {
    this.client = mqtt.connect('ws://localhost:9001');

    this.client.on('connect', () => {
      console.log('Connected to MQTT broker');
    });

    this.client.on('error', (err) => {
      console.error('MQTT Connection error:', err);
    });

    this.client.on('message', (topic, message) => {
      console.log(`Message received on ${topic}: ${message.toString()}`);
    });
  }

  subscribe(topic) {
    if (this.client) {
      this.client.subscribe(topic, (err) => {
        if (!err) {
          console.log(`Subscribed to topic: ${topic}`);
        } else {
          console.error('Subscription error:', err);
        }
      });
    }
  }

  publish(topic, message) {
    if (this.client) {
      this.client.publish(topic, message);
    }
  }
}

export default new MqttService();
